### DISCUSSION TITLE
How about a define-type-alias?

### DISCUSSION THREAD

**Discussion by bon** [coalton-lang/coalton#375] [Ideas and Proposals]
...or some other moral equivalent of the Haskell [type](https://wiki.haskell.org/Type#Type_and_newtype) statement for introducing a synonym. It would be very helpful as an organisational tool.
Judging by a couple of comments in the thih example, I'm not the first person to think this.

**Comment by stylewarning** [comment]
I agree this could be useful. I know other communities are divided on the matter (type aliases make code hard to debug). I've just been using "newtypes" (i.e., ADTs with one constructor), which have no run-time penalty in release-mode, but have a penalty in terms of characters written.

**Reply by bon** [reply]
You mean something like this? 
```lisp
(defpackage :newtypes
  (:use :coalton :coalton-library))

(in-package :newtypes)

(coalton-toplevel

  (define-type Foobar String)

  ;; (declare fact (Integer -> Foobar -> Integer))
  (define (fact n f)
    (if (== n 0)
        1
        (fact (- n 1) f))))
```
A call such as `(fact 4 "superfluous")` fails to unify if the declaration is uncommented, saying `"superfluous"` is of type `String` not `Foobar`.

Without the declaration it takes any type in the second slot.

**Reply by stylewarning** [reply]
No:

```
(define-type Foo (Foo String))
```

You still need to construct and match against `Foo`, but it has no runtime cost.

**Reply by bon** [reply]
Thanks! A bit wordy but it works.

**Reply by ornamentist** [reply]
Would this work for a type alias to a function type, e.g. `Relation` as `:a -> :a -> Boolean`?

**Reply by stylewarning** [reply]
> Would this work for a type alias to a function type, e.g. `Relation` as `:a -> :a -> Boolean`?

You can't hide the `:a` even with a type alias. You'd have to have `(Relation :a)`. The above would still work.

**Reply by ornamentist** [reply]
Thanks - I should have said Relation :a above.

**Reply by ornamentist** [reply]
Would a macro work to avoid the match/deconstruct steps, particularly for a "simplifying" type alias to a more complex type definition? Some kind of (nested) `define-alias` macro?

**Reply by stylewarning** [reply]
> Would a macro work to avoid the match/deconstruct steps, particularly for a "simplifying" type alias to a more complex type definition? Some kind of (nested) `define-alias` macro?

I think having first-class support for something like `define-type-alias` would be best. I don't think macros would dig us out if that a hole.
